### ISSUE TITLE
Selector is expected to be initialized with a function, CSS selector

### ISSUE THREAD

**Issue by Meemaw** [testing-library/testcafe-testing-library#38]
After upgrading to `testcafe` v1.6.0 it seems that `@testing-library/testcafe` selectors started failing:

I can see this repo is still on `v1.5.0` so it probably is not compatible with `v1.6.0` yet?

```js
import { getByPlaceholderText } from '@testing-library/testcafe';

  const searchInput = getByPlaceholderText( 'Type...');
  await t
    .expect(searchInput.exists)
    .ok('Search input exists')
    .typeText(searchInput, 'randomText');
```

Throws:
<img width="724" alt="Screen Shot 2019-10-18 at 11 57 02 AM" src="https://user-images.githubusercontent.com/8524109/67085053-6d285f80-f19e-11e9-9753-8cdaf1ac262f.png">

Error points to the `.typeText` line.

 
### Your Environment details:
 
* testcafe version:                   1.6.0
* node.js version:                    v8.16.0
* browser name and version: HeadlessChrome 68.0.3440

P.S.
I have also opened this issue in testcafe repo: https://github.com/DevExpress/testcafe/issues/4400

**Comment by benmonro** [comment]
Thanks for reporting this.  what version of TCTL are you on?

**Comment by benmonro** [comment]
hmm, ok yeah I just update (https://github.com/testing-library/testcafe-testing-library/pull/39) and it is passing all tests.  let me try to repro locally

**Comment by Meemaw** [comment]
Great, if you'll have issues reproducing I will get you a reproducible example tomorrow -- don't spend too much time on it.

**Comment by benmonro** [comment]
yeah was able to run your example locally w/ no problem.  reproducible sample will be much appreciated. 🙌

**Comment by benmonro** [comment]
@Meemaw any update on this?

**Comment by benmonro** [comment]
@Meemaw closing this for now since I haven't heard from you.  if you need to reopen please do.

**Comment by rssfrncsrenishaw** [comment]
I'm getting the same error in 1.8.1

**Comment by rssfrncsrenishaw** [comment]
@Meemaw  did you ever solve the issue?

**Comment by Meemaw** [comment]
No, I didn't really have too much to dig into it so we just stopped using this library.

**Comment by benmonro** [comment]
I'm happy to look into this if you can provide repro @rssfrncsrenishaw

**Comment by kkateli** [comment]
I have the same error:
Selector is expected to be initialized with a function, CSS selector string, another Selector, node snapshot or a Promise returned by a Selector, but object was passed.

testCafe: 1.8.2
@testing-library/testcafe: 4.0.0

**Comment by benmonro** [comment]
@kkateli any chance you can create a reproducible example?  I was never able to...

**Comment by kkateli** [comment]
@benmonro I found out that testcafe has to be a **devDependency** to make the error go away. Now it works fine now. Thanks for your quick response.

**Comment by benmonro** [comment]
Whoa that's odd.  I'll try to see if I can repro it that way.  Thank you for getting back.  Hope you enjoy the library and feel free to post any other feedback

**Comment by cbayer97** [comment]
I'm having the same issue with roughly the following setup:
```typescript
import { t } from 'testcafe'

export default class WebElement {
  constructor(private selector: Selector) {}

  public async click() {
    await t.click(this.selector)
  }

  //...

}
```

```typescript
import WebElement from './WebElement'
import { getByText } from '@testing-library/testcafe'

fixture('Foo')

test('Bar', async (t) => {
  button = new WebElement(getByText('button-text'))
  button.click()
})
```

I actually do have the TCTL as well as testcafe itself imported as a devDependency, but still get that error:
```
1) Action "selector" argument error:
      
Selector is expected to be initialized with a function, CSS selector string, another Selector, node snapshot or a Promise returned by a Selector, but object was passed.
```

## Environment details:
* "node": "12.14.0"
* "@testing-library/testcafe": "4.0.0"
* "testcafe": "1.8.2",
*  Browser: Chrome 80.0.3987.149 / macOS 10.15.3

**Comment by benmonro** [comment]
@cbayer97  curious if you can try @kkateli 's fix above and ensure its a dev dep...

**Comment by benmonro** [comment]
@cbayer97  ok interesting I'll try to reproduce on my end.  Ignore my last comment noticed you mentioned the devdep

**Comment by benmonro** [comment]
@cbayer97 so I attempted to recreate the issue on this PR: https://github.com/testing-library/testcafe-testing-library/pull/130.  It appears to work for me locally as well in CI, but I'm doing it in javascript.  I did have to change the constructor of WebElement to initialize the selector, not sure if that matters in typescript.

```jsx
import { t } from 'testcafe';

export default class WebElement {
    constructor(selector) {
        this.selector = selector; // <-- I added this
    }

    async click() {
        await t.click(this.selector)
    }

}
```

**Comment by benmonro** [comment]
@cbayer97  let me know if that works for you.

**Comment by benmonro** [comment]
@cbayer97 i just tried to re-create your problem in a create-react-app using the typescript template.  your code you had above works fine.  One thing to note, did you add the following to your `.testcaferc.json` file:

```json
"clientScripts": [
    "./node_modules/@testing-library/dom/dist/@testing-library/dom.umd.js"
  ]
```
You can see my full working example here:
https://github.com/benmonro/testcafe-testing-library-typescript-example

if you're still having issues if you wouldn't mind trying to push a reproducible example to a repo I can take a look, but using the code you've provided works for me...

**Comment by benmonro** [comment]
@cbayer97 did this work for you?

**Comment by cbayer97** [comment]
Sorry for the late response @benmonro. We use a [runner](https://github.com/DevExpress/testcafe/blob/master/docs/articles/documentation/using-testcafe/programming-interface/runner.md) to start and configure our testcafe setup and thus do not have a `.testcaferc.json`. The `dom.umd.js` script, however, is imported in the runner. Also your suggested solution with the constructor did not work for me. I have checked out your example repo and tried to simulate the same environment that we use. I was able to break it down to the typescript version. In your example you used `~3.7.2` whereas we use `3.8.3`. Updating the minor version causes the tests to err, even in your example.

**Comment by benmonro** [comment]
Ah ok thank you this is helpful I'll see if I can repro on my end

**Comment by Meemaw** [comment]
Reproducer:

```sh
git clone https://github.com/meemaw/insight testcafe-bug-1
cd testcafe-bug-1
yarn install
cd frontend/try
yarn dev
```

Another shell:

```sh
cd frontend/try
yarn test:e2e
```

The setup is quite  complex here (monorepo with Lerna).

**Comment by TeaBough** [comment]
I have the same error:
Selector is expected to be initialized with a function, CSS selector string, another Selector, node snapshot or a Promise returned by a Selector, but object was passed.

testCafe: 1.7.0
@testing-library/testcafe: 4.0.0

it works with testCafe: 1.6.1

**Comment by benmonro** [comment]
Hey @Meemaw really appreciate the repro on that.  looking into it this morning.  A couple notes so far.  When I add t.debug() in your test before any of the selectors are executed, and go into the chrome dev tools, I'm seeing this error:

![image](https://user-images.githubusercontent.com/399236/80918401-90b98e80-8d19-11ea-9929-61f81aa86fd7.png)

I also noticed that if I mimic what testcafe testing library does directly in your test it also works:

```typescript
    .typeText(Selector(() => {
       return window["TestingLibraryDom"].getByPlaceholderText(document.body, 'First name');
    }), 'Joe')
```

So my suspicion right now is around the fact that it's a monorepo.  will keep digging though and report back...

**Comment by benmonro** [comment]
ah ha! I think I found the issue.  it looks like you are on testcafe@1.8.4 and @testing-library/testcafe is using testcafe@1.8.3.  I published an alpha build that moves testcafe into peer dependencies to avoid this issue and tested it against my fork of your repo and it seems to work.  Can you verify it on your end and I will cut it as a release.

https://github.com/Meemaw/Insight/compare/master...benmonro:master


also not related, but in case you're interested, testcafe now supports modules in the config file so you can get rid of the `../../node_modules` by doing this instead in .testcaferc.json

```diff
  "clientScripts": [
-    "../../node_modules/@testing-library/dom/dist/@testing-library/dom.umd.js"
+    {"module":"@testing-library/dom/dist/@testing-library/dom.umd.js"}
  ]
```

**Comment by Meemaw** [comment]
I can confirm that version fixes the issue.

**Comment by benmonro** [comment]
Fixed in 4.0.1
